### PR TITLE
Limiting refactorings to the appropriate scope

### PR DESCRIPTION
**Pull request by mattwynne**
@kevinrutherford @tooky let's work on / discuss this in here.

### CODE REVIEW

**Comment by kevinrutherford** [comment]
What if the given scope is a block? What if the temp is actually defined outside of that block?

**Comment by tooky** [comment]
I think we'll need line and column, for one line blocks.

**Comment by chrismdp** [comment]
We will need to worry about method scoping, though. We won't necessarily want to change the temp name for the whole tree, if the name is reused in different methods.

**Comment by chrismdp** [comment]
Feels like we need some way of delimiting traversals, both up and down the tree. Something like:

```
surrounding_scope() #=> sexp giving the starting point of the current scope for line and column
```

and

```
traverse_with_no_subscopes(sexpr) #=> doesn't traverse into newly created scopes
```

(obviously with better names)

For the traverse, feels like we could create a little set of filters/operations to be run on each sexp within a tree, with fail fast if needed.

**Comment by kevinrutherford** [comment]
> ```
> traverse_with_no_subscopes(sexpr) #=> doesn't traverse into newly created scopes
> ```

Yes. In reek _every_ traverse omitted nested class, module and method
definitions within the traversed scope.
But blocks and nested blocks are a different case, I think.

**Comment by mattwynne** [comment]
One test at a time lads. Somebody make the spec I added pass, then we can worry about blocks on the next iteration.

File: spec/raffle/refactorings/rename_temp_spec.rb [hunk 1]
```diff
@@ -2,16 +2,87 @@
...
         fred = 45
         june = fred
       end
+
+      def bar
+        fred = "captain"
+      end
     }
-    output = refactor(input, 'fred', 'billy')
+    output = refactor(input, 'fred', 'billy', [3,0])
```
Comment (mattwynne) [inline spec/raffle/refactorings/rename_temp_spec.rb]: Note that these args mean: `old_name, new_name, location`
where `location` means [`row`, `col`]

The idea is that the location should be a 'hit' from anywhere inside a scope.

**Comment by chrismdp** [comment]
After some more experimentation: wow - this gets complicated real fast. We need to store the last variable definition, and push it onto a stack somehow such that it gets removed when we finish traversing the current scope. Before finishing this up it might be worth moving to decorating sexps with functionality, or building our own tree.

**Comment by kevinrutherford** [comment]
In reek I tried both, and decorating was by far the most successful.
See https://github.com/troessner/reek/blob/master/lib/reek/source/tree_dresser.rb
for the decorator code.

**Comment by mattwynne** [comment]
It's getting a little bit tiring passing sexps around, but I'm actually quite happy with the simplicity of using pure functions for now.

**Comment by mattwynne** [comment]
@chrismdp I've made the whitespace pending spec into a pull request so it's more visible. Once @jimweirich gets back from the mountains we can figure that one out. For now we'll have to keep all the examples flat.

**Comment by chrismdp** [comment]
Agreed. Liking the flow of using pull requests to track discussion of projected features.

File: lib/raffle/refactorings/navigates_trees.rb [hunk 2]
```diff
@@ -9,12 +9,31 @@ def walk(sexp, &block)
...
+      def transform_within_scope(current_sexp, scope_sexp, &block)
+        return current_sexp unless current_sexp.respond_to?(:map)
+        current_sexp.map do |child_sexp|
+          if (child_sexp == scope_sexp)
+            transform(child_sexp, &block)
+          else
+            transform_within_scope(child_sexp, scope_sexp, &block)
+          end
+        end
+      end
```
Comment (chrismdp) [inline lib/raffle/refactorings/navigates_trees.rb:21]: I don't much like this method, but appears the easiest way to propagate and correctly reset state.

Any ideas for refactoring this?

**Comment by chrismdp** [comment]
Ok, I've made these cases work and all the specs pass. It occurs to me that we don't yet check for other types of scope changes such as for/while loops and begin/end blocks.

**Comment by chrismdp** [comment]
It's also pretty ugly code. Looking forward to some refactoring attempts :)

File: lib/raffle/refactorings/rename_temp.rb [hunk 3]
```diff
@@ -1,12 +1,38 @@
...
+        position_sexp = sexp_for_position(starting_sexp, line, column)
+        scope_sexp = find_containing_scope_in(starting_sexp, position_sexp)
+        unless_scope_defines_variable_again = lambda do |new_scope|
+          return true if new_scope == scope_sexp
+          return false if find_first(new_scope) do |sexp|
+            block_parameter?(sexp, original_name)
+          end
+          true
+        end
+        transform_within_scope(starting_sexp, scope_sexp, unless_scope_defines_variable_again) do |sexp|
```
Comment (mattwynne) [inline lib/raffle/refactorings/rename_temp.rb:27]: Yeah wow, this beauty needs some refactoring love!

Comment (mattwynne) [inline lib/raffle/refactorings/rename_temp.rb:27]: How about starting by creating a Raffle::Position class?

File: lib/raffle/refactorings/rename_temp.rb [hunk 4]
```diff
@@ -1,12 +1,30 @@
...
+
+      def call(starting_sexp, original_name, new_name, line_and_column)
+        scope_sexp = containing_scope_for_position(starting_sexp, Position.new(line_and_column))
+        unless_scope_defines_variable_again = lambda do |new_scope|
+          not defining_block_parameter?(new_scope, original_name)
+        end
+        transform_within_scope(starting_sexp, scope_sexp, unless_scope_defines_variable_again) do |sexp|
           next unless ident?(sexp, original_name)
           [:@ident, new_name, sexp[2]]
         end
```
Comment (chrismdp) [inline lib/raffle/refactorings/rename_temp.rb:30]: Ok - had a good go at the method - this is a lot neater. Still need better names surrounding the lambda, but hopefully it's readable.

We've now got some quite powerful methods to do with scope: we can find the scope for a position, we can traverse and only change a particular scope, and we can choose whether to enter an inner scope based on a block.

We could go super-functional at this point and generate the transform method from a two 'strategy' lambdas, but I'm not sure how to do recursive lambda generation in Ruby... :) Perhaps we'd be better off sticking plain strategy composites to handle the scoping.

Any reason not to merge this into master soon?

**Comment by chrismdp** [comment]
Can someone please read, understand and review this code before merging? Thanks.

**Comment by tooky** [comment]
I went through it all and happy I understand it, so I merged it.

**Comment by chrismdp** [comment]
Thanks, Steve. Feels like we should get most code reviewed on a pull request like this before merging it.

## 

Chris Parsons
http://twitter.com/chrismdp
http://pa.rsons.org
(from my phone)

**Pull request merged**